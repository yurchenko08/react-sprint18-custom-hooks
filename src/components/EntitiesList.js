import { useState, useEffect } from "react";

function EntitiesList({ onEntityDetailsClick, propsToDisplay, fetchMethod }) {
  const [entities, setEntities] = useState(null);

  useEffect(() => {
    fetchMethod().then((res) => {
      setEntities(res);
      console.log("Users: ", res);
    });
  }, [fetchMethod]);

  return entities ? (
    <ul>
      {entities.map((entity) => (
        <li key={entity.id}>
          <button onClick={() => onEntityDetailsClick(entity.id)}>ðŸ‘€</button>
          {Object.entries(propsToDisplay).map((entry) => (
            <span key={entity[entry[0]]}>
              {" "}
              {entry[1]}: <strong>{entity[entry[0]]}</strong>
            </span>
          ))}
        </li>
      ))}
    </ul>
  ) : (
    <p className="center">Loading...</p>
  );
}

export default EntitiesList;
