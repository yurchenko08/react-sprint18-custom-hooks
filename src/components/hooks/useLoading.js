import { useState, useEffect } from 'react';
export const useLoading = (fetchMethod, params) => {
  const [entity, setEntity] = useState(null);

  useEffect(() => {
    fetchMethod(params).then((res) => {
      setEntity(res);
      console.log('Users: ', res);
    });
  }, [fetchMethod, params]);

  return entity;
};
