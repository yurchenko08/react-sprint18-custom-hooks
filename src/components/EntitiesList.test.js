import renderer from "react-test-renderer";
import EntitiesList from "../components/EntitiesList";
import TestRenderer from "react-test-renderer";
import { setImmediate } from "timers";
import * as useLoadingContext from "./hooks/useLoading";
import { render } from "@testing-library/react";

const { act } = TestRenderer;

describe("EntitiesList", () => {
  it("renders correctly with data", async () => {
    const wrapper = renderer.create(
      <EntitiesList
        fetchMethod={() =>
          new Promise((resolve) => resolve([{ id: 101, name: "John" }]))
        }
        propsToDisplay={{ name: "Name" }}
      />
    );
    await act(() => new Promise(setImmediate));

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it("renders loading without data", async () => {
    const wrapper = renderer.create(
      <EntitiesList
        fetchMethod={() =>
          new Promise((resolve) => resolve([{ id: 101, name: "John" }]))
        }
        propsToDisplay={{ name: "Name" }}
      />
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it("checks if useLoading was called", () => {
    const mySpy = jest
      .spyOn(useLoadingContext, "useLoading")
      .mockImplementation(() => []);
    render(
      <EntitiesList
        fetchMethod={() => new Promise(() => [])}
        propsToDisplay={{}}
      />
    );

    expect(mySpy).toHaveBeenCalledTimes(1);
  });
});
