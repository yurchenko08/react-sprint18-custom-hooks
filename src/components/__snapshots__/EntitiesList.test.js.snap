// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntitiesList renders correctly with data 1`] = `
<ul>
  <li>
    <button
      onClick={[Function]}
    >
      ðŸ‘€
    </button>
    <span>
       
      Name
      : 
      <strong>
        John
      </strong>
    </span>
  </li>
</ul>
`;

exports[`EntitiesList renders loading without data 1`] = `
<p
  className="center"
>
  Loading...
</p>
`;
